airflow/models/dagcode.py

```diff
@@ -0,0 +1,213 @@
+                orm_dag_code.last_updated = timezone.utcnow()
+                orm_dag_code.source_code = DagCode._read_code(orm_dag_code.fileloc)
+                session.update(orm_dag_code)
+
+    @classmethod
+    @provide_session
+    def remove_deleted_code(cls, alive_dag_filelocs: List[str], session=None):
+        """Deletes code not included in alive_dag_filelocs.
+
+        :param alive_dag_filelocs: file paths of alive DAGs
+        :param session: ORM Session
+        """
+        alive_fileloc_hashes = [
+            cls.dag_fileloc_hash(fileloc) for fileloc in alive_dag_filelocs]
+
+        log.debug("Deleting code from %s table ", cls.__tablename__)
+
+        session.execute(
+            session.query(cls).filter(
+                and_(cls.fileloc_hash.notin_(alive_fileloc_hashes),
+                     cls.fileloc.notin_(alive_dag_filelocs))).delete())
+
+    @classmethod
+    @provide_session
+    def has_dag(cls, fileloc: str, session=None) -> bool:
+        """Checks a file exist in dag_code table.
+
+        :param fileloc: the file to check
+        :param session: ORM Session
+        """
+        fileloc_hash = cls.dag_fileloc_hash(fileloc)
+        return session.query(exists().where(cls.fileloc_hash == fileloc_hash))\
+            .scalar()
+
+    @classmethod
+    def get_code_by_fileloc(cls, fileloc: str) -> str:
+        """Returns source code for a given fileloc.
+
+        :param fileloc: file path of a DAG
+        :return: source code as string
+        """
+        return DagCode(fileloc).code()
+
+    def code(self) -> str:
+        """Returns source code for this DagCode object.
+
+        :return: source code as string
+        """
+        if conf.getboolean('core', 'store_dag_code', fallback=False):
+            return self._get_code_from_db()
+        else:
+            return self._get_code_from_file()
+
+    def _get_code_from_file(self):
+        with open_maybe_zipped(self.fileloc, 'r') as f:
+            code = f.read()
+        return code
+
+    @provide_session
+    def _get_code_from_db(self, session=None):
+        dag_code = session.query(DagCode) \
+            .filter(DagCode.fileloc_hash == self.fileloc_hash) \
+            .first()
+        if not dag_code:
+            raise DagCodeNotFound()
+        else:
+            code = dag_code.source_code
+        return code
+
+    @staticmethod
+    def dag_fileloc_hash(full_filepath: str) -> int:
+        """"Hashing file location for indexing.
+
+        :param full_filepath: full filepath of DAG file
+        :return: hashed full_filepath
+        """
+        # Hashing is needed because the length of fileloc is 2000 as an Airflow convention,
+        # which is over the limit of indexing.
+        import hashlib
+        # Only 7 bytes because MySQL BigInteger can hold only 8 bytes (signed).
+        return struct.unpack('>Q', hashlib.sha1(
+            full_filepath.encode('utf-8')).digest()[-8:])[0] >> 8
```

Comment:
@anitakar Did you mean for this construct to ignore the least-significant byte?

Location: `/home/ash/airflow/dags/example.py`
SHA1 (hex) `0xx78288e229ef15e32a7a32e1fd123d9fc60b5eae0`
fileloc_hash: `58867689281598954`
fileloc_hash hex: `d123d9fc60b5ea`

i.e. it ignores/removes the `e0` from the end.

Reply:
In all honesty I wanted to use all 8 bytes of mysql biginteger.
I used another construct before but it was not working with python 2, so I have changed it to use python struct. But I should have used signed long long instead of unsigned long long.
When I have noted it the code was too close to releasing and 7 bytes is enough anyway (https://cloud.google.com/composer/docs/release-notes).

Reply:
Figured it was something like that, just wanted to check if there was a deeper reason. Thanks.